### PR TITLE
Implemented Subroutines EIP2315

### PR DESCRIPTION
**Pull request by adria0**
Backport of https://github.com/openethereum/openethereum/commit/234a2877be91a9beecde89e4177558c6c6689231 + https://github.com/openethereum/openethereum/commit/58fde9834d61b160675560fca6aad4368f9ad746

### CODE REVIEW

File: ethcore/src/trace/types/error.rs [hunk 1]
```diff
@@ -82,8 +90,9 @@ impl fmt::Display for Error {
 			BadInstruction => "Bad instruction",
 			StackUnderflow => "Stack underflow",
 			OutOfStack => "Out of stack",
-			BuiltIn => "Built-in failed",
-			Wasm => "Wasm runtime error",
+			SubStackUnderflow => "Subroutine stack underflow",
+			OutOfSubStack => "Subroutine stack overflow",			BuiltIn => "Built-in failed",
+			InvalidSubEntry => "Invalid subroutine entry",			Wasm => "Wasm runtime error",
```
Comment (rakita) [inline ethcore/src/trace/types/error.rs:95]: Is this intended to be on same line without line break?

Comment (adria0) [reply ethcore/src/trace/types/error.rs:95]: Oops! I'll fix it

File: ethcore/src/trace/types/error.rs [hunk 2]
```diff
@@ -57,7 +63,9 @@ impl<'a> From<&'a VmError> for Error {
 			VmError::BadInstruction { .. } => Error::BadInstruction,
 			VmError::StackUnderflow { .. } => Error::StackUnderflow,
 			VmError::OutOfStack { .. } => Error::OutOfStack,
-			VmError::BuiltIn { .. } => Error::BuiltIn,
+			VmError::SubStackUnderflow { .. } => Error::SubStackUnderflow,
+			VmError::OutOfSubStack { .. } => Error::OutOfSubStack,
+			VmError::InvalidSubEntry { .. } => Error::InvalidSubEntry,			VmError::BuiltIn { .. } => Error::BuiltIn,
```
Comment (rakita) [inline ethcore/src/trace/types/error.rs:68]: one more here, sorry i miss it first time.

Comment (adria0) [reply ethcore/src/trace/types/error.rs:68]:
```suggestion
			VmError::InvalidSubEntry { .. } => Error::InvalidSubEntry,
			VmError::BuiltIn { .. } => Error::BuiltIn,
```

File: ethcore/evm/src/interpreter/mod.rs [hunk 3]
```diff
@@ -62,6 +62,10 @@ const TWO_POW_96: U256 = U256([0, 0x100000000, 0, 0]); //0x1 00000000 00000000 0
 const TWO_POW_224: U256 = U256([0, 0, 0, 0x100000000]); //0x1 00000000 00000000 00000000 00000000 00000000 00000000 00000000
 const TWO_POW_248: U256 = U256([0, 0, 0, 0x100000000000000]); //0x1 00000000 00000000 00000000 00000000 00000000 00000000 00000000 000000
 
+/// Maximal subroutine stack size as specified in
```
Comment (vorot93) [inline ethcore/evm/src/interpreter/mod.rs:65]:
```suggestion
/// Maximum subroutine stack size as specified in
```

File: ethcore/evm/src/interpreter/mod.rs [hunk 4]
```diff
@@ -420,6 +431,21 @@ impl<Cost: CostType> Interpreter<Cost> {
 				};
 				self.reader.position = pos;
 			},
+			InstructionResult::JumpToSubroutine(position) => {
+				if self.valid_subroutine_destinations.is_none() {
+					self.valid_subroutine_destinations = Some(self.cache.jump_and_sub_destinations(&self.params.code_hash, &self.reader.code).1);
+				}
```
Comment (vorot93) [inline ethcore/evm/src/interpreter/mod.rs:437]:
```suggestion
				self.valid_subroutine_destinations.get_or_insert_with(|| self.cache.jump_and_sub_destinations(&self.params.code_hash, &self.reader.code).1);
```

File: ethcore/evm/src/interpreter/mod.rs [hunk 5]
```diff
@@ -444,15 +470,17 @@ impl<Cost: CostType> Interpreter<Cost> {
...
+		if (instruction == DELEGATECALL && !schedule.have_delegate_call) ||
+			(instruction == CREATE2 && !schedule.have_create2) ||
+			(instruction == STATICCALL && !schedule.have_static_call) ||
+			((instruction == RETURNDATACOPY || instruction == RETURNDATASIZE) && !schedule.have_return_data) ||
+			(instruction == REVERT && !schedule.have_revert) ||
+			((instruction == SHL || instruction == SHR || instruction == SAR) && !schedule.have_bitwise_shifting) ||
+			(instruction == EXTCODEHASH && !schedule.have_extcodehash) ||
+			(instruction == CHAINID && !schedule.have_chain_id) ||
+			(instruction == SELFBALANCE && !schedule.have_selfbalance) ||
+			((instruction == BEGINSUB || instruction == JUMPSUB || instruction == RETURNSUB) && !schedule.have_subs)
```
Comment (vorot93) [inline ethcore/evm/src/interpreter/mod.rs:483]: One reason to have a single big reformatting/restyling commit before everything else.

**Pull request merged**